### ISSUE TITLE
Thellier GUI Measurement Quality Branch

### ISSUE THREAD

**Issue by Caoimhinmg** [PmagPy/PmagPy#170]
I have over the course of the last few months created a local branch of PmagPy where I worked on making Thellier GUI capable of marking measurement data bad by right clicking on the measurement record in the list on the left of the GUI. This makes Demag GUI and Thellier GUI more similar in the design of their user interaction which simplifies the UI, prevents the deletion of unwanted measurement data which is the current method to solve the lack of this feature, and allows the user to abstract another aspect of data analysis which is one of the goals of the GUIs. This new feature with the exception of marking NRM steps bad, which I have marked by a warning dialog and basically disallowed, is done. I have pushed a moment ago this branch of PmagPy named thellier_meas_flags and I hope that @ltauxe, @ronshaar, and @moonshoes87 could take a quick look at this branch and ensure it is more or less stable and works mostly as expected before I merge it into master. So I'll wait for a few responses on the function of this feature bellow before I do the merge which I hope to be able to do in the next couple of weeks before our next release. I have already checked that it is merged properly with master up to now (as far as I can tell), and that it passes all of the unit tests we currently have.

**Comment by ltauxe** [comment]
it is not so easy to delete a measurement for thellier data as you really need to delete at least a pair and you have to be careful of pTRM checks up the line and vector subtraction issues.  did you do that?  
lisa

**Comment by Caoimhinmg** [comment]
Yeah I noticed the problems with removing a measurement in a Thellier experiment are much greater than in a Demag experiment. So here's the basic overview:

----------------------------

- Z or I step marked bad:
    - if there is another Z or I step of matching treatment use it instead and recalculate parameters (basically restarts a segment of the GUI, hope I didn't miss anything here because this is where things could go wrong) these are the IZZ, IIZ, ZZI, ZII cases where one of the duplicates is bad.
    - if there are no matching Z or I steps of the same treatment you have no choice but to throw out that treatment step and remove it from calculations even though there might be a different Z or I step still marked good. Data is reinitialized as if this treatment step is just non-existent. These are the ZI or IZ cases where one is bad. So basically how to handle only a single I or Z for a treatment step the solution being that there is no treatment step if you only have a single I or Z you need both so if there is one it just may as well not be there.
- PTRM checks, Tail checks, Additivity checks:
    - These are all just single data point measurements so I can just remove the bad points from the dataset and recalculate the parameters pertaining to differences between data points.
- NRM
    - Freak out and tell the user not to do this. Thellier GUI regularly accesses the first Z step as the NRM because that's where it puts it during data read in. This becomes a bit more troubling as later I need to keep that data but not treat it as the NRM to normalize things with. This also raises the problem of what do you do if the user marks all NRMs bad? I might look at this more later, but for now just implementing this for the other steps should be a large bonus for the users. By the way though I still let the user do this, but it does not prevent the NRM from being used for normalizing things and just removes it from PCA there is a really long warning message to tell the user this is kinda a bad idea. I can fully block it before merging if there are objections to this.

----------------------------

There are some troubles with this (mostly around the recalculation bits), but it is the most simplistic and intuitive approach I could think of that works without too much implementation mess. That said I might have missed some of the places where data is stored in Thellier GUI as it has data all over and such a major feature is doubtless to have fallout. Which is why I am hoping to test this rigorously on the separate live branch before merging. Also it would be good to have someone with better knowledge of Thellier GUI's data structures see if I caught all the places I need to recalculate after removing a point. This could mostly be done by checking the recalculation of statistics after marking things good or bad in the GUI and just looking at how the functions mark_meas_bad, mark_meas_good, and on_right_click_listctrl work none of which are long.

**Comment by moonshoes87** [comment]
@Caoimhinmg, are you still willing/able to implement this?  @ltauxe and I were just meeting and agreed that this approach makes sense.  If you don't have the time, I may try to simply implement @ltauxe's #277 request (plotting measurements that were already marked as bad in the measurement file with an x)

**Comment by Caoimhinmg** [comment]
It already works to some degree in the branch I left. If I remember right there were just a few calculation issues to hammer out because the functions that take in data to do statistics on them are not prepared to handle bad measurements. Particularly because there are so many different kinds of measurements in a standard paleointensity study at least compared to a paleodirection study. So that said this branch already does what Lisa requests in #277 on all plots and exactly as Demag GUI does it.

I could defiantly go about finishing this. I'll try to look this weekend and see what work there is left for this branch. Though it would also be good if @ltauxe could look at the way this branch is handling measurements marked bad and if the way it's changing the calculations is logical/correct, because I'm no great paleointensity expert like she is and can't do the basic "oh that makes sense" calculations in my head to check my work here. With that we should be able to map out exactly what is left needing to be done for the full integration of this feature.

**Comment by ltauxe** [comment]
I finally got a chance to look into this, but of course the branch is still in python2 so doesn't run on my computer and the two projects have diverged since you wrote this (my bad - really sorry).  But your logic above looks sound to me.  The only trouble I can foresee is that some folks start their plots at high temperature (say 500C!) so somehow we should be able to ignore the NRM if it isn't present and not complain (use the first zero field step for the normalization).  But that is not really your problem.

**Comment by moonshoes87** [comment]
@Caoimhinmg, I know it's been a long time, but is there any chance that you would have some time to port this over to python/MagIC 3?  Let me know if I can help, or maybe take this on entirely if you are too busy?
